backtesting/backtesting.py

```diff
@@ -1514,3 +1525,68 @@ def plot(self, *, results: pd.Series = None, filename=None, plot_width=None,
             reverse_indicators=reverse_indicators,
             show_legend=show_legend,
             open_browser=open_browser)
+
+    def optimize_skopt(self,
+                       maximize: Union[str, Callable[[pd.Series], float]] = 'SQN',
+                       max_tries: Union[int, float] = 100,
+                       **kwargs) -> Union[pd.Series, Tuple[pd.Series, pd.Series]]:
+        """
+        Optimize strategy parameters to an optimal combination using
+        scikit-optimize. Returns result `pd.Series` of
+        the best run.
+
+        `maximize` is a string key from the
+        `backtesting.backtesting.Backtest.run`-returned results series,
+        or a function that accepts this series object and returns a number;
+        the higher the better. By default, the method maximizes
+        Van Tharp's [System Quality Number](https://google.com/search?q=System+Quality+Number).
+
+        Additional keyword arguments represent strategy arguments with
+        list-like collections of possible values. For example, the following
+        code finds and returns the "best" of the 7 admissible (of the
+        9 possible) parameter combinations:
+
+            backtest.optimize(sma1=[5, 10, 15], sma2=[10, 20, 40],
+                              constraint=lambda p: p.sma1 < p.sma2)
+
+        .. TODO::
+            Improve multiprocessing with joblib.parallel .
+        """
+
+        if not kwargs:
+            raise ValueError('Need some strategy parameters to optimize')
+
+        maximize_key = None
+        if isinstance(maximize, str):
+            maximize_key = str(maximize)
+            stats = self._results if self._results is not None else self.run()
+            if maximize not in stats:
+                raise ValueError('`maximize`, if str, must match a key in pd.Series '
+                                 'result of backtest.run()')
+
+            def maximize(stats: pd.Series, _key=maximize):
+                return stats[_key]
+
+        elif not callable(maximize):
+            raise TypeError('`maximize` must be str (a field of backtest.run() result '
+                            'Series) or a function that accepts result Series '
+                            'and returns a number; the higher the better')
+
+        dimensions = [Integer(name='n1', low=min(kwargs.get('n1')), high=max(kwargs.get('n1'))),
+                      Integer(name='n2', low=min(kwargs.get('n2')), high=max(kwargs.get('n2')))]
```

Comment:
Integer and min/max logic is fine for integer input, but we'll need to cover reals and categoricals also. I guess you can walk over `kwargs` values lists and wrap them as `np.array`. Then if `dtype.kind` is `'i'`, create an Integer dimension, if `'f'` (float), [a Real dimension, and a Categorical dimension](https://scikit-optimize.github.io/stable/modules/classes.html#module-skopt.space.space) in all other cases?